File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> a;
        queue<pair<int,TreeNode*>> q;
        if(!root)
        {
            return {};
        }
        q.push({0,root});
        while(!q.empty())
        {
            TreeNode *temp = q.front().second;
            int level = q.front().first;
            if(level == a.size())
            {
                a.push_back(vector<int>());
            }
            a[level].push_back(temp->val);
            q.pop();
            if(temp->left)
            {
                q.push({level+1,temp->left});
            }
            if(temp->right)
            {
                q.push({level+1,temp->right});
            }
        }
        return a;
        
        
    }
};

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func rec(root *TreeNode) int{
    if root == nil{
        return 0
    }
    left := rec(root.Left) + 1
    right := rec(root.Right) + 1
    return max(left,right)
}

func isBalanced(root *TreeNode) bool {
    // base condition
    if root == nil{
        return true
    }
    left := rec(root.Left)
    right := rec(root.Right)
    if math.Abs(float64(left - right)) > 1{
        return false
    }
    leftbal := isBalanced(root.Left)
    rightbal := isBalanced(root.Right)

    if !leftbal || !rightbal{
        return false
    }else{
        return true
    }
}
